src/Main.hs

```diff
@@ -14,15 +14,143 @@
+
+main :: IO()
+main = writePng "output.png" $ generateImage Main.trace sceneWidth sceneHeight
+
+minWithEmpty :: [Double] -> Double
+minWithEmpty [] = -1
+minWithEmpty list = minimum list
+
+trace x y 
+    | isNothing intersection = white
+    | otherwise = getColorFromIntersection intersection
+    where intersection = closestIntersection (generateRay camera sceneWidth sceneHeight x y) objects
+
+getColorFromIntersection (Just ( _ , Object _ (Material color))) = color
+
+-- Generating rays, assuming distance to the image is 1 unit
+generateRay :: Camera -> Int -> Int -> Int -> Int -> Ray
+generateRay camera width height x y = 
+    let 
+         -- Vector from camera to lookAt point
+        eyeVector = buildEyeVector camera 
+        -- Vector in local right direction
+        rightVector = normalize (eyeVector `cross` Vector 0 1 0)
+        upVector = normalize (rightVector `cross` eyeVector)
+        -- Halves are taken to make right angles
+        halfFov = fovInRadians camera
+        -- This aspect ratio will be used, but are not the width and height of the camera
+        aspectRatio = h / w
+        halfWidth = tan halfFov
+        halfHeight = aspectRatio * halfWidth
+        cameraWidth = halfWidth * 2
+        cameraHeight = halfHeight * 2
+        pixelWidth = cameraWidth / (w - 1)
+        pixelHeight = cameraHeight / (h - 1)
+        scaledX = ((fromIntegral x * pixelWidth) - halfWidth)  `scalarMult` rightVector
+        scaledY = ((fromIntegral y * pixelHeight) - halfHeight) `scalarMult`  upVector 
+        orientation = normalize $ eyeVector `add` scaledX `add` scaledY
+    in  Ray (getPosition camera) orientation
+    where w = fromIntegral width
+          h = fromIntegral height
+
+buildEyeVector :: Camera -> Vector
+buildEyeVector (Camera _ location lookAt) = normalize (lookAt `sub` location)
+
+fovInRadians :: Camera -> Double
+fovInRadians (Camera fov _ _ ) = pi * (fov / 2) / 180
+
+getPosition :: Camera -> Vector
+getPosition (Camera _ position _ ) = position
+
+closestIntersection :: Ray -> [Object] -> Maybe (Double, Object)
+closestIntersection ray objects
+    | null intersections = Nothing
+    | otherwise = Just $ minimumBy tupleMinimum (map fromJust intersections)
+    where intersections = filter isJust $ map (intersects ray) objects
+
+tupleMinimum :: (Double, Object) -> (Double,Object) -> Ordering
+tupleMinimum x y
+    | fst x < fst y = LT
+    | fst x > fst y = GT
+    | otherwise = EQ
+
+-- Minimum distance intersection
+intersects :: Ray -> Object -> Maybe (Double, Object)
+intersects (Ray origin direction) (Object (Sphere center radius) material) =
+    let
+        l = origin `sub` center
+        a = direction `dot` direction
+        b = 2 * (direction `dot` l)
+        c =  (l `dot` l) - radius^2
+        listOfRoots = roots a b c
+    in
+        case listOfRoots of
+            [] -> Nothing
+            otherwise -> Just (minimum listOfRoots, Object (Sphere center radius) material)
```

Comment:
could simplify this a touch by not recreating a new sphere
http://stackoverflow.com/questions/1153465/what-does-the-symbol-mean-in-reference-to-lists-in-haskell